### ISSUE TITLE
Generalizing to different image sizes

### ISSUE THREAD

**Issue by ykaplanq** [gengshan-y/high-res-stereo#13]
I have two questions: a theoretical one, and a question about the practical training process .
The original purpose of the network is to be able to process any image size thanks to pyramid pooling blocks. However, it seems that during training all input data is cropped (and/or padded) to a specific size of 768x512.  Since pyramid pooling kernel size depends on image size, I wonder how the network (specifically, pyramid pooling layers in feature extraction and decoder) will be able to generalize to any image size?
Regarding the practical training process - I tried to train the network from scratch by following the procedure described in the README.
When running the evaluation on Middlebury validation images using --testres 1, I get an average error of 4.6. Running the same evaluation using the pretrained weights you provided, I got an average error of 1.92.
I tried several things to improve accuracy, including:
- Fine-tuning on the Middlebury and Kitti training + validation data
- Using a max disparity of 768 during training time
- Several suggestions given in response to issues on the GitHub forum
The lowest average error I’ve managed to get is 3.
Could you please describe the full training process to recreate the model provided on the site? Did I miss something in the data sets, training scripts, their parameters or the training process itself?

**Comment by gengshan-y** [comment]
> Since pyramid pooling kernel size depends on image size, I wonder how the network (specifically, pyramid pooling layers in feature extraction and decoder) will be able to generalize to any image size?

The motivation of volumetric pyramid pooling (VPP) is closer to that of pyramid pooling module in [PSPNet](https://arxiv.org/abs/1612.01105): efficiently aggregate global and semi-global information of the input tensor by pooling. The way to make it generalize is to **resize** the inputs at a wide range and crop to a fixed size when training, so that intuitively the model not only see high-res local structures but also low-res global structures.

> When running the evaluation on Middlebury validation images using --testres 1, I get an average error of 4.6. Running the same evaluation using the pretrained weights you provided, I got an average error of 1.92.The lowest average error I’ve managed to get is 3.

The provided model is trained on train+validation set of Middlebury, and that's why you see a gap. Without including the validation set, 3 seems to be a reasonable error.

**Comment by ykaplanq** [comment]
Thank you. As a follow up question:
As you suggested, I used the Middlebury validation data set during training, but still could not reproduce your results. 
So far, I’ve done the entire training process with:
1.	Max disparity set to 384 and scale set to 1.
2.	Max disparity set to 384 and scale set to 2.
3.	Max disparity set to 768 and scale set to 2.

In all cases, we followed the initial training process with fine-tuning on the full KITTI and Middlebury datasets.
None of the models created this way give a validation error of less than 2.6. Surprisingly, training with max disparity set to 768 gave a particularly high error of 5.3 on the Middlebury validation data, even though those images were included in the training data.
Can you tell us what other changes to make in the script, and which max disparity to use, to reproduce your results?

**Comment by gengshan-y** [comment]
We used the 3rd setting to train the final model,
> Max disparity set to 768 and scale set to 2.

It seems a bit strange to get such a large error (5.3). I guess there are a few things to check: 
1) Did you use all the five datasets and a batch size of 24, as in the training script? 
2) The training could be unstable. Did you try evaluating the intermediate models during training? Is the error actually going down or is it always around 5?
